File: internal/deployment/package_graph_test.go
package deployment

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"strings"
	"testing"

	"github.com/helmrdotdev/helmr/internal/jsoncanon"
)

func TestPackageGraphMatchesSharedGoldenFixture(t *testing.T) {
	fixture := loadContractFixture(t)
	graph, err := ParsePackageGraph([]byte(fixture.PackageGraph.Canonical))
	if err != nil {
		t.Fatal(err)
	}
	canonical, err := CanonicalPackageGraph(graph)
	if err != nil {
		t.Fatal(err)
	}
	if string(canonical) != fixture.PackageGraph.Canonical {
		t.Fatalf("canonical package graph = %q, want %q", canonical, fixture.PackageGraph.Canonical)
	}
	digest := sha256.Sum256(canonical)
	if hex.EncodeToString(digest[:]) != fixture.PackageGraph.DigestHex {
		t.Fatalf("package graph digest = %x, want %s", digest, fixture.PackageGraph.DigestHex)
	}
}

func TestPackageGraphRejectsSharedMutations(t *testing.T) {
	fixture := loadContractFixture(t)
	for _, test := range fixture.PackageGraphRejections {
		t.Run(test.Name, func(t *testing.T) {
			value := packageGraphFixtureValue(t, fixture.PackageGraph.Canonical)
			locals := value["localPackages"].([]any)
			registries := value["registryPackages"].([]any)
			resolutions := value["resolutions"].([]any)
			root := locals[0].(map[string]any)
			local := locals[1].(map[string]any)
			registry := registries[0].(map[string]any)
			switch test.Mutation {
			case "missing_format_version":
				delete(value, "formatVersion")
			case "unknown_root_member":
				value["unknown"] = true
			case "unknown_local_member":
				local["unknown"] = true
			case "unknown_registry_member":
				registry["unknown"] = true
			case "unknown_resolution_member":
				resolutions[0].(map[string]any)["unknown"] = true
			case "unknown_endpoint_member":
				resolutions[0].(map[string]any)["from"].(map[string]any)["unknown"] = true
			case "root_not_first":
				locals[0], locals[1] = locals[1], locals[0]
			case "duplicate_root":
				value["localPackages"] = append(locals, cloneJSONValue(t, root))
			case "local_order":
				value["localPackages"] = append(locals, localFixture("packages/alpha", nil, nil))
			case "overlapping_local_path":
				value["localPackages"] = append(locals, localFixture("packages/shared/nested", nil, nil))
			case "non_adjacent_overlapping_local_path":
				value["localPackages"] = []any{
					locals[0],
					localFixture("a", nil, nil),
					localFixture("a-", nil, nil),
					localFixture("a/b", nil, nil),
					locals[1],
				}
			case "reserved_local_path":
				setFixtureLocalPath(t, value, "packages/shared", "helmr/shared")
			case "absolute_local_path":
				setFixtureLocalPath(t, value, "packages/shared", "/packages/shared")
			case "local_view_key":
				local["viewKey"] = strings.Repeat("0", 64)
			case "root_view_key":
				root["viewKey"] = strings.Repeat("0", 64)
			case "ambiguous_local_name":
				local["name"] = "app"
			case "registry_order":
				registries[0], registries[1] = registries[1], registries[0]
			case "duplicate_registry_path":
				registries[1].(map[string]any)["installPath"] = registry["installPath"]
			case "reserved_registry_path":
				registry["installPath"] = ".helmr/package"
			case "registry_integrity":
				registry["integrity"] = "sha512-invalid"
			case "registry_name":
				registry["name"] = "Invalid"
			case "empty_registry_version":
				registry["version"] = ""
			case "resolution_order":
				resolutions[0], resolutions[1] = resolutions[1], resolutions[0]
			case "resolution_relationship":
				resolutions[0].(map[string]any)["relationship"] = "development"
			case "resolution_dependency":
				resolutions[0].(map[string]any)["dependency"] = "Invalid"
			case "missing_from_node":
				resolutions[0].(map[string]any)["from"].(map[string]any)["path"] = "packages/missing"
			case "missing_to_node":
				resolutions[0].(map[string]any)["to"].(map[string]any)["installPath"] = "missing"
			case "unnamed_local_target":
				local["name"] = nil
			case "local_target_alias":
				resolutions[1].(map[string]any)["dependency"] = "alias"
			case "mixed_endpoint_shape":
				resolutions[0].(map[string]any)["from"].(map[string]any)["installPath"] = "zod"
			case "oversized_path_component":
				setFixtureLocalPath(t, value, "packages/shared", "packages/"+strings.Repeat("a", 256))
			case "oversized_mounted_path":
				setFixtureLocalPath(t, value, "packages/shared", pathWithLength(4096-len(programMountPath)-1))
			default:
				t.Fatalf("unknown fixture mutation %q", test.Mutation)
			}
			requirePackageGraphRejection(t, value)
		})
	}
}

func TestPackageGraphRejectsMissingNullAndDuplicateMembers(t *testing.T) {
	fixture := loadContractFixture(t)
	for _, member := range []string{"formatVersion", "localPackages", "registryPackages", "resolutions"} {
		for _, mode := range []string{"missing", "null"} {
			t.Run(mode+"_root_"+member, func(t *testing.T) {
				value := packageGraphFixtureValue(t, fixture.PackageGraph.Canonical)
				if mode == "missing" {
					delete(value, member)
				} else {
					value[member] = nil
				}
				requirePackageGraphRejection(t, value)
			})
		}
	}

	type memberSet struct {
		array    string
		position int
		members  []string
	}
	for _, object := range []memberSet{
		{array: "localPackages", position: 1, members: []string{"manifestDigest", "name", "path", "version", "viewKey"}},
		{array: "registryPackages", position: 0, members: []string{"installPath", "integrity", "name", "version"}},
		{array: "resolutions", position: 0, members: []string{"dependency", "from", "relationship", "to"}},
	} {
		for _, member := range object.members {
			t.Run("missing_"+object.array+"_"+member, func(t *testing.T) {
				value := packageGraphFixtureValue(t, fixture.PackageGraph.Canonical)
				entry := value[object.array].([]any)[object.position].(map[string]any)
				delete(entry, member)
				requirePackageGraphRejection(t, value)
			})
			if object.array != "localPackages" || (member != "name" && member != "version") {
				t.Run("null_"+object.array+"_"+member, func(t *testing.T) {
					value := packageGraphFixtureValue(t, fixture.PackageGraph.Canonical)
					entry := value[object.array].([]any)[object.position].(map[string]any)
					entry[member] = nil
					requirePackageGraphRejection(t, value)
				})
			}
		}
	}
	for _, endpointName := range []string{"from", "to"} {
		for _, member := range []string{"kind", "path"} {
			for _, mode := range []string{"missing", "null"} {
				t.Run(mode+"_"+endpointName+"_"+member, func(t *testing.T) {
					value := packageGraphFixtureValue(t, fixture.PackageGraph.Canonical)
					endpoint := value["resolutions"].([]any)[1].(map[string]any)[endpointName].(map[string]any)
					if mode == "missing" {
						delete(endpoint, member)
					} else {
						endpoint[member] = nil
					}
					requirePackageGraphRejection(t, value)
				})
			}
		}
	}

	for _, test := range fixture.PackageGraphRawRejections {
		t.Run(test.Name, func(t *testing.T) {
			raw := fixture.PackageGraph.Canonical
			switch test.Mutation {
			case "duplicate_root_member":
				raw = strings.Replace(raw, `"formatVersion":0`, `"formatVersion":0,"formatVersion":0`, 1)
			case "duplicate_local_member":
				raw = strings.Replace(raw, `"manifestDigest":"sha256:000`, `"manifestDigest":"sha256:`+strings.Repeat("0", 64)+`","manifestDigest":"sha256:000`, 1)
			case "duplicate_endpoint_member":
				raw = strings.Replace(raw, `"from":{"kind":"local"`, `"from":{"kind":"local","kind":"local"`, 1)
			default:
				t.Fatalf("unknown raw fixture mutation %q", test.Mutation)
			}
			if _, err := ParsePackageGraph([]byte(raw)); err == nil || !strings.Contains(err.Error(), "duplicate object name") {
				t.Fatalf("ParsePackageGraph error = %v, want duplicate object name", err)
			}
		})
	}
}

func TestPackageGraphAcceptsRootOnlyCyclesAndOpaqueVersions(t *testing.T) {
	fixture := loadContractFixture(t)
	rootOnly := PackageGraph{
		FormatVersion:    PackageGraphFormatVersion,
		LocalPackages:    []LocalPackage{{ManifestDigest: "sha256:" + strings.Repeat("0", 64), Path: "."}},
		RegistryPackages: []RegistryPackage{},
		Resolutions:      []PackageResolution{},
	}
	canonical, err := CanonicalPackageGraph(rootOnly)
	if err != nil {
		t.Fatal(err)
	}
	if string(canonical) != fixture.PackageGraph.RootOnlyCanonical {
		t.Fatalf("canonical root-only graph = %q, want %q", canonical, fixture.PackageGraph.RootOnlyCanonical)
	}
	digest := sha256.Sum256(canonical)
	if hex.EncodeToString(digest[:]) != fixture.PackageGraph.RootOnlyDigestHex {
		t.Fatalf("root-only graph digest = %x, want %s", digest, fixture.PackageGraph.RootOnlyDigestHex)
	}
	if _, err := ParsePackageGraph(canonical); err != nil {
		t.Fatal(err)
	}

	graph, err := ParsePackageGraph([]byte(fixture.PackageGraph.Canonical))
	if err != nil {
		t.Fatal(err)
	}
	localVersion := "file:opaque"
	graph.LocalPackages[1].Version = &localVersion
	graph.RegistryPackages[0].Version = "link:opaque"
	root := graph.LocalPackages[0]
	local := graph.LocalPackages[1]
	graph.Resolutions = []PackageResolution{
		{
			Dependency:   *local.Name,
			From:         PackageEndpoint{Kind: PackageKindLocal, Path: &root.Path},
			Relationship: PackageRelationshipProduction,
			To:           PackageEndpoint{Kind: PackageKindLocal, Path: &local.Path},
		},
		{
			Dependency:   *root.Name,
			From:         PackageEndpoint{Kind: PackageKindLocal, Path: &local.Path},
			Relationship: PackageRelationshipProduction,
			To:           PackageEndpoint{Kind: PackageKindLocal, Path: &root.Path},
		},
	}
	if _, err := CanonicalPackageGraph(graph); err != nil {
		t.Fatal(err)
	}
}

func TestPackageGraphRejectsRegistryToLocalResolution(t *testing.T) {
	fixture := loadContractFixture(t)
	graph, err := ParsePackageGraph([]byte(fixture.PackageGraph.Canonical))
	if err != nil {
		t.Fatal(err)
	}
	local := graph.LocalPackages[1]
	registry := graph.RegistryPackages[0]
	graph.Resolutions = []PackageResolution{{
		Dependency: *local.Name,
		From: PackageEndpoint{
			InstallPath: &registry.InstallPath,
			Kind:        PackageKindRegistry,
		},
		Relationship: PackageRelationshipProduction,
		To: PackageEndpoint{
			Kind: PackageKindLocal,
			Path: &local.Path,
		},
	}}
	if err := ValidatePackageGraph(graph); err == nil ||
		!strings.Contains(err.Error(), "registry-to-local resolution is unsupported") {
		t.Fatalf("ValidatePackageGraph error = %v", err)
	}
}

func TestPackageGraphUsesUnsignedUTF8PathOrderWithoutNormalization(t *testing.T) {
	paths := []string{"packages/e\u0301", "packages/é", "packages/\ue000", "packages/😀"}
	locals := []LocalPackage{{ManifestDigest: "sha256:" + strings.Repeat("0", 64), Path: "."}}
	for position, path := range paths {
		viewKey := localPackageViewKey(path)
		locals = append(locals, LocalPackage{
			ManifestDigest: "sha256:" + strings.Repeat(string(rune('1'+position)), 64),
			Path:           path,
			ViewKey:        &viewKey,
		})
	}
	graph := PackageGraph{
		FormatVersion:    PackageGraphFormatVersion,
		LocalPackages:    locals,
		RegistryPackages: []RegistryPackage{},
		Resolutions:      []PackageResolution{},
	}
	canonical, err := CanonicalPackageGraph(graph)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := ParsePackageGraph(canonical); err != nil {
		t.Fatal(err)
	}
}

func TestPackageGraphPathAndScalarBounds(t *testing.T) {
	if err := validatePackagePath(pathWithLength(4096-len(programMountPath)-2), programMountPath, true); err != nil {
		t.Fatal(err)
	}
	if err := validatePackagePath(pathWithLength(4096-len(programMountPath)-1), programMountPath, true); err == nil {
		t.Fatal("validatePackagePath accepted a mounted path over 4096 bytes")
	}
	if err := validatePackagePath("packages/"+strings.Repeat("a", 255), programMountPath, true); err != nil {
		t.Fatal(err)
	}
	if err := validatePackagePath("packages/"+strings.Repeat("a", 256), programMountPath, true); err == nil {
		t.Fatal("validatePackagePath accepted a component over 255 bytes")
	}
	if err := validatePackageName(strings.Repeat("a", 214)); err != nil {
		t.Fatal(err)
	}
	if err := validatePackageName(strings.Repeat("a", 215)); err == nil {
		t.Fatal("validatePackageName accepted a name over 214 bytes")
	}
	if err := validatePackageVersion(strings.Repeat("v", 255)); err != nil {
		t.Fatal(err)
	}
	if err := validatePackageVersion(strings.Repeat("v", 256)); err == nil {
		t.Fatal("validatePackageVersion accepted a version over 255 bytes")
	}
}

func TestPackageGraphRejectsNonCanonicalAndBoundedInput(t *testing.T) {
	fixture := loadContractFixture(t)
	if _, err := ParsePackageGraph([]byte(" " + fixture.PackageGraph.Canonical)); err == nil {
		t.Fatal("ParsePackageGraph accepted non-canonical input")
	}
	if _, err := ParsePackageGraph(nil); err == nil {
		t.Fatal("ParsePackageGraph accepted empty input")
	}
	if _, err := ParsePackageGraph(make([]byte, maxProgramFileSizeBytes+1)); err == nil {
		t.Fatal("ParsePackageGraph accepted oversized input")
	}
}

func packageGraphFixtureValue(t *testing.T, canonical string) map[string]any {
	t.Helper()
	var value map[string]any
	if err := json.Unmarshal([]byte(canonical), &value); err != nil {
		t.Fatal(err)
	}
	return value
}

func requirePackageGraphRejection(t *testing.T, value map[string]any) {
	t.Helper()
	raw, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := ParsePackageGraph(canonical); err == nil {
		t.Fatal("ParsePackageGraph returned nil error")
	}
}

func setFixtureLocalPath(t *testing.T, value map[string]any, oldPath, newPath string) {
	t.Helper()
	locals := value["localPackages"].([]any)
	for _, item := range locals {
		local := item.(map[string]any)
		if local["path"] == oldPath {
			local["path"] = newPath
			local["viewKey"] = localPackageViewKey(newPath)
		}
	}
	for _, item := range value["resolutions"].([]any) {
		resolution := item.(map[string]any)
		for _, endpointName := range []string{"from", "to"} {
			endpoint := resolution[endpointName].(map[string]any)
			if endpoint["kind"] == "local" && endpoint["path"] == oldPath {
				endpoint["path"] = newPath
			}
		}
	}
}

func localFixture(path string, name, version *string) map[string]any {
	return map[string]any{
		"manifestDigest": "sha256:" + strings.Repeat("2", 64),
		"name":           name,
		"path":           path,
		"version":        version,
		"viewKey":        localPackageViewKey(path),
	}
}

func cloneJSONValue(t *testing.T, value any) any {
	t.Helper()
	raw, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	var cloned any
	if err := json.Unmarshal(raw, &cloned); err != nil {
		t.Fatal(err)
	}
	return cloned
}

func pathWithLength(length int) string {
	var builder strings.Builder
	for builder.Len() < length {
		if builder.Len() > 0 {
			builder.WriteByte('/')
		}
		remaining := length - builder.Len()
		if remaining > maxPackagePathComponent {
			remaining = maxPackagePathComponent
		}
		builder.WriteString(strings.Repeat("z", remaining))
	}
	return builder.String()
}

File: sdk/typescript/src/internal/package-graph.test.ts
import { describe, expect, test } from "bun:test"
import { createHash } from "node:crypto"
import { readFile } from "node:fs/promises"
import { resolve } from "node:path"

import { canonicalizeJson } from "./jsoncanon"
import {
  canonicalPackageGraph,
  PACKAGE_GRAPH_FORMAT_VERSION,
  parsePackageGraph,
  type PackageGraph,
} from "./package-graph"

interface PackageGraphFixture {
  readonly packageGraph: {
    readonly canonical: string
    readonly digestHex: string
    readonly rootOnlyCanonical: string
    readonly rootOnlyDigestHex: string
  }
  readonly packageGraphRejections: readonly {
    readonly name: string
    readonly mutation: string
  }[]
  readonly packageGraphRawRejections: readonly {
    readonly name: string
    readonly mutation: string
  }[]
}

const decoder = new TextDecoder()

describe("deployment package graph", async () => {
  const fixture = await loadFixture()

  test("matches the shared canonical fixture", () => {
    const graph = parsePackageGraph(fixture.packageGraph.canonical)
    const canonical = canonicalPackageGraph(graph)
    expect(decoder.decode(canonical)).toBe(fixture.packageGraph.canonical)
    expect(createHash("sha256").update(canonical).digest("hex")).toBe(fixture.packageGraph.digestHex)
  })

  test("rejects the shared invalid mutations", () => {
    for (const item of fixture.packageGraphRejections) {
      const value = packageGraphFixtureValue(fixture.packageGraph.canonical)
      const locals = value["localPackages"] as Array<Record<string, unknown>>
      const registries = value["registryPackages"] as Array<Record<string, unknown>>
      const resolutions = value["resolutions"] as Array<Record<string, unknown>>
      const root = locals[0]!
      const local = locals[1]!
      const registry = registries[0]!
      switch (item.mutation) {
        case "missing_format_version":
          delete value["formatVersion"]
          break
        case "unknown_root_member":
          value["unknown"] = true
          break
        case "unknown_local_member":
          local["unknown"] = true
          break
        case "unknown_registry_member":
          registry["unknown"] = true
          break
        case "unknown_resolution_member":
          resolutions[0]!["unknown"] = true
          break
        case "unknown_endpoint_member":
          ;(resolutions[0]!["from"] as Record<string, unknown>)["unknown"] = true
          break
        case "root_not_first":
          ;[locals[0], locals[1]] = [locals[1]!, locals[0]!]
          break
        case "duplicate_root":
          locals.push(structuredClone(root))
          break
        case "local_order":
          locals.push(localFixture("packages/alpha", null, null))
          break
        case "overlapping_local_path":
          locals.push(localFixture("packages/shared/nested", null, null))
          break
        case "non_adjacent_overlapping_local_path":
          value["localPackages"] = [
            locals[0],
            localFixture("a", null, null),
            localFixture("a-", null, null),
            localFixture("a/b", null, null),
            locals[1],
          ]
          break
        case "reserved_local_path":
          setFixtureLocalPath(value, "packages/shared", "helmr/shared")
          break
        case "absolute_local_path":
          setFixtureLocalPath(value, "packages/shared", "/packages/shared")
          break
        case "local_view_key":
          local["viewKey"] = "0".repeat(64)
          break
        case "root_view_key":
          root["viewKey"] = "0".repeat(64)
          break
        case "ambiguous_local_name":
          local["name"] = "app"
          break
        case "registry_order":
          ;[registries[0], registries[1]] = [registries[1]!, registries[0]!]
          break
        case "duplicate_registry_path":
          registries[1]!["installPath"] = registry["installPath"]
          break
        case "reserved_registry_path":
          registry["installPath"] = ".helmr/package"
          break
        case "registry_integrity":
          registry["integrity"] = "sha512-invalid"
          break
        case "registry_name":
          registry["name"] = "Invalid"
          break
        case "empty_registry_version":
          registry["version"] = ""
          break
        case "resolution_order":
          ;[resolutions[0], resolutions[1]] = [resolutions[1]!, resolutions[0]!]
          break
        case "resolution_relationship":
          resolutions[0]!["relationship"] = "development"
          break
        case "resolution_dependency":
          resolutions[0]!["dependency"] = "Invalid"
          break
        case "missing_from_node":
          ;(resolutions[0]!["from"] as Record<string, unknown>)["path"] = "packages/missing"
          break
        case "missing_to_node":
          ;(resolutions[0]!["to"] as Record<string, unknown>)["installPath"] = "missing"
          break
        case "unnamed_local_target":
          local["name"] = null
          break
        case "local_target_alias":
          resolutions[1]!["dependency"] = "alias"
          break
        case "mixed_endpoint_shape":
          ;(resolutions[0]!["from"] as Record<string, unknown>)["installPath"] = "zod"
          break
        case "oversized_path_component":
          setFixtureLocalPath(value, "packages/shared", `packages/${"a".repeat(256)}`)
          break
        case "oversized_mounted_path":
          setFixtureLocalPath(value, "packages/shared", pathWithLength(4096 - programMountPath.length - 1))
          break
        default:
          throw new Error(`unknown fixture mutation ${item.mutation}`)
      }
      expectPackageGraphRejection(value, item.name)
    }
  })

  test("rejects missing, null, and duplicate members", () => {
    for (const member of ["formatVersion", "localPackages", "registryPackages", "resolutions"]) {
      for (const mode of ["missing", "null"] as const) {
        const value = packageGraphFixtureValue(fixture.packageGraph.canonical)
        if (mode === "missing") delete value[member]
        else value[member] = null
        expectPackageGraphRejection(value, `${mode} root ${member}`)
      }
    }

    const objects = [
      { array: "localPackages", position: 1, members: ["manifestDigest", "name", "path", "version", "viewKey"] },
      { array: "registryPackages", position: 0, members: ["installPath", "integrity", "name", "version"] },
      { array: "resolutions", position: 0, members: ["dependency", "from", "relationship", "to"] },
    ] as const
    for (const object of objects) {
      for (const member of object.members) {
        const missing = packageGraphFixtureValue(fixture.packageGraph.canonical)
        const missingEntry = (missing[object.array] as Array<Record<string, unknown>>)[object.position]!
        delete missingEntry[member]
        expectPackageGraphRejection(missing, `missing ${object.array}.${member}`)

        if (object.array !== "localPackages" || (member !== "name" && member !== "version")) {
          const nullValue = packageGraphFixtureValue(fixture.packageGraph.canonical)
          const nullEntry = (nullValue[object.array] as Array<Record<string, unknown>>)[object.position]!
          nullEntry[member] = null
          expectPackageGraphRejection(nullValue, `null ${object.array}.${member}`)
        }
      }
    }

    for (const endpointName of ["from", "to"]) {
      for (const member of ["kind", "path"]) {
        for (const mode of ["missing", "null"] as const) {
          const value = packageGraphFixtureValue(fixture.packageGraph.canonical)
          const resolution = (value["resolutions"] as Array<Record<string, unknown>>)[1]!
          const endpoint = resolution[endpointName] as Record<string, unknown>
          if (mode === "missing") delete endpoint[member]
          else endpoint[member] = null
          expectPackageGraphRejection(value, `${mode} ${endpointName}.${member}`)
        }
      }
    }

    for (const item of fixture.packageGraphRawRejections) {
      let raw = fixture.packageGraph.canonical
      switch (item.mutation) {
        case "duplicate_root_member":
          raw = raw.replace('"formatVersion":0', '"formatVersion":0,"formatVersion":0')
          break
        case "duplicate_local_member":
          raw = raw.replace(
            '"manifestDigest":"sha256:000',
            `"manifestDigest":"sha256:${"0".repeat(64)}","manifestDigest":"sha256:000`,
          )
          break
        case "duplicate_endpoint_member":
          raw = raw.replace('"from":{"kind":"local"', '"from":{"kind":"local","kind":"local"')
          break
        default:
          throw new Error(`unknown raw fixture mutation ${item.mutation}`)
      }
      expect(() => parsePackageGraph(raw), item.name).toThrow(/duplicate/)
    }
  })

  test("accepts root-only graphs, cycles, and opaque version strings", () => {
    const rootOnly: PackageGraph = {
      formatVersion: PACKAGE_GRAPH_FORMAT_VERSION,
      localPackages: [
        {
          manifestDigest: `sha256:${"0".repeat(64)}`,
          name: null,
          path: ".",
          version: null,
          viewKey: null,
        },
      ],
      registryPackages: [],
      resolutions: [],
    }
    const canonical = canonicalPackageGraph(rootOnly)
    expect(decoder.decode(canonical)).toBe(fixture.packageGraph.rootOnlyCanonical)
    expect(createHash("sha256").update(canonical).digest("hex")).toBe(
      fixture.packageGraph.rootOnlyDigestHex,
    )
    expect(() => parsePackageGraph(canonical)).not.toThrow()

    const value = packageGraphFixtureValue(fixture.packageGraph.canonical)
    ;(value["localPackages"] as Array<Record<string, unknown>>)[1]!["version"] = "file:opaque"
    ;(value["registryPackages"] as Array<Record<string, unknown>>)[0]!["version"] = "link:opaque"
    expect(() => parsePackageGraph(canonicalizeJson(JSON.stringify(value)))).not.toThrow()
  })

  test("uses unsigned UTF-8 path order without Unicode normalization", () => {
    const paths = ["packages/e\u0301", "packages/é", "packages/\ue000", "packages/😀"]
    const graph: PackageGraph = {
      formatVersion: PACKAGE_GRAPH_FORMAT_VERSION,
      localPackages: [
        {
          manifestDigest: `sha256:${"0".repeat(64)}`,
          name: null,
          path: ".",
          version: null,
          viewKey: null,
        },
        ...paths.map((path, position) => ({
          manifestDigest: `sha256:${String(position + 1).repeat(64)}`,
          name: null,
          path,
          version: null,
          viewKey: fixtureViewKey(path),
        })),
      ],
      registryPackages: [],
      resolutions: [],
    }
    expect(() => parsePackageGraph(canonicalPackageGraph(graph))).not.toThrow()
  })

  test("enforces scalar, component, and mounted path bounds", () => {
    const value = packageGraphFixtureValue(fixture.packageGraph.canonical)
    const root = (value["localPackages"] as Array<Record<string, unknown>>)[0]!
    root["name"] = "a".repeat(214)
    root["version"] = "v".repeat(255)
    expect(() => parsePackageGraph(canonicalizeJson(JSON.stringify(value)))).not.toThrow()

    root["name"] = "a".repeat(215)
    expectPackageGraphRejection(value, "oversized name")
    root["name"] = "app"
    root["version"] = "v".repeat(256)
    expectPackageGraphRejection(value, "oversized version")

    const acceptedPath = pathWithLength(4096 - programMountPath.length - 2)
    setFixtureLocalPath(value, "packages/shared", acceptedPath)
    root["version"] = null
    expect(() => parsePackageGraph(canonicalizeJson(JSON.stringify(value)))).not.toThrow()
  })

  test("rejects non-canonical, empty, and oversized input", () => {
    expect(() => parsePackageGraph(` ${fixture.packageGraph.canonical}`)).toThrow(/canonical/)
    expect(() => parsePackageGraph(new Uint8Array())).toThrow(/size/)
    expect(() => parsePackageGraph(new Uint8Array(16777217))).toThrow(/size/)
  })
})

const programMountPath = "/opt/helmr/program"

async function loadFixture(): Promise<PackageGraphFixture> {
  const path = resolve(import.meta.dir, "../../../../fixtures/contracts/deployment-v0/golden.json")
  return JSON.parse(await readFile(path, "utf8")) as PackageGraphFixture
}

function packageGraphFixtureValue(canonical: string): Record<string, unknown> {
  return JSON.parse(canonical) as Record<string, unknown>
}

function expectPackageGraphRejection(value: Record<string, unknown>, label: string): void {
  const canonical = canonicalizeJson(JSON.stringify(value))
  expect(() => parsePackageGraph(canonical), label).toThrow()
}

function setFixtureLocalPath(
  value: Record<string, unknown>,
  oldPath: string,
  newPath: string,
): void {
  for (const local of value["localPackages"] as Array<Record<string, unknown>>) {
    if (local["path"] === oldPath) {
      local["path"] = newPath
      local["viewKey"] = fixtureViewKey(newPath)
    }
  }
  for (const resolution of value["resolutions"] as Array<Record<string, unknown>>) {
    for (const endpointName of ["from", "to"]) {
      const endpoint = resolution[endpointName] as Record<string, unknown>
      if (endpoint["kind"] === "local" && endpoint["path"] === oldPath) {
        endpoint["path"] = newPath
      }
    }
  }
}

function localFixture(path: string, name: string | null, version: string | null): Record<string, unknown> {
  return {
    manifestDigest: `sha256:${"2".repeat(64)}`,
    name,
    path,
    version,
    viewKey: fixtureViewKey(path),
  }
}

function fixtureViewKey(path: string): string {
  return createHash("sha256")
    .update("helmr.local-package-view.v0")
    .update(new Uint8Array([0]))
    .update(path)
    .digest("hex")
}

function pathWithLength(length: number): string {
  let path = ""
  while (path.length < length) {
    if (path.length > 0) path += "/"
    path += "z".repeat(Math.min(255, length - path.length))
  }
  return path
}
